src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabSCMSource.java

```diff
@@ -318,20 +328,18 @@ protected void retrieve(SCMSourceCriteria criteria, @NonNull SCMHeadObserver obs
                                 )
                         );
                         String originOwner = m.getAuthor().getUsername();
-                        // Origin project name will always the same as the source project name
-                        String originProjectPath = projectPath;
+                        // This is a hack to get the namespace from the path with namespace
+                        String originProjectPath = gitLabApi.getProjectApi().getProject(m.getProjectId()).getPathWithNamespace();
```

Comment:
This call is only necessary if `fork` is true.
So could be rewritten as such:

```java
                        boolean fork = !m.getSourceProjectId().equals(m.getTargetProjectId());
                        String originProjectPath = projectPath;
                        if (fork) {
                            originProjectPath = gitLabApi.getProjectApi().getProject(m.getProjectId()).getPathWithNamespace();
                        }
```

Reply:
Perhaps with multiple forks you could have little memory map outside loop to avoid repeated calls to the same project?

Reply:
Nice catch. Fixed in 8449ace.

Reply:
@baymac I would be concerned about gitlab.com's 600 API calls per minute: https://gitlab.com/gitlab-org/gitlab-ce/issues/41308#note_87693185

Reply:
Hmm! Can you describe a situation where it cause problems?

Reply:
> Perhaps with multiple forks you could have little memory map outside loop to avoid repeated calls to the same project?

Fixed in f7d4851.

Reply:
> Hmm! Can you describe a situation where it cause problems?

Sure :)
API limits is a way to ensure stability this post is properly better at explaining it: https://apisyouwonthate.com/blog/what-is-api-rate-limiting-all-about

As soon as we hit the 600 requests within a minute we would have to wait till RateLimit is reset.
This reset limit can is returned on gitlab.com in the headers.
But for self-hosted GitLab it currently not returned: https://gitlab.com/gitlab-org/gitlab-ce/issues/41309#note_65515282

Currently, we don't have any code that handles the back-off and retries when we hit the API limit.

Reply:
Shouldn't this be handled inside gitlab-api?